### PR TITLE
fix(core): dispatch loadEntry, pass entities to widget

### PR DESCRIPTION
**Pull request by d4rekanguok**
Hi folks, I'm so glad to see many updates to this project in the last few months. Thanks for the hardwork!

**Summary**

In the latest version, the method `loadEntry` that was passed down to widgets no longer work (I've been relying on this API for a few custom widgets in production). I digged around the repo to see what changed & send this PR just to facilitate communication.

This is how it worked before:

```js
async componentDidMount() {
  const { loadEntry } = this.props
  const results = await loadEntry(collectionName, slug)
  // results.payload.data. etc etc</p>
}
```

I found 3 issues:

- In EditorControl, it seems that `loadEntry` wasn't properly dispatched here, resulting in failed call
https://github.com/netlify/netlify-cms/blob/ab7ba435baf07862e7adbfaf0f6cd9950af7ad52/packages/netlify-cms-core/src/components/Editor/EditorControlPane/EditorControl.js#L282-L304

- In `actions/entries/loadEntry`, `createDraftFromEntry(loadedEntry)` refreshes the widget component and cause an infinite loop. I patched this by adding a flag, but there's probably better ways to handle this. https://github.com/netlify/netlify-cms/blob/ab7ba435baf07862e7adbfaf0f6cd9950af7ad52/packages/netlify-cms-core/src/actions/entries.ts#L343-L347

- Also in the same code, in the previous version the action's payload is returned, but no longer the case in the latest version:
```js
    await waitForMediaLibraryToLoad(dispatch, getState());
    dispatch(entryLoading(collection, slug));
    return backend
      .getEntry(getState(), collection, slug)
      .then((loadedEntry: EntryValue) => {
        return dispatch(entryLoaded(collection, loadedEntry)); // return payload
      })
``` 
I think it is more correct to not return data from action (however this is currently still the case with the `query` method). But then, widgets can't access the returned data. I ended up passing `state.entries.entities` to widget. this works but now handling data in widget is a bit more acrobatic. Here's what that [looks like](https://github.com/d4rekanguok/netlify-cms-widgets/blob/707169003d8eed68289b2abcac71f564e6dafa4f/packages/widget-file-relation/src/control.tsx#L18-L50):

```js
  public async componentDidMount() {
    const { loadEntry, field } = this.props
    
    // ✂ snipped some code

    loadEntry(collection, file)
  }

  public componentDidUpdate() {
    const { field, entities } = this.props

    // ✂ snipped some code

    const entity = entities.get(`${collection}.${file}`)
    if (entity) {
      const data = entity.get('data').get(fieldName)
      const options = data.map(option => ({
        value: option.get(fieldId),
        label: option.get(fieldDisplay),
      }))
      this.setState({ options })
    }
  }
```

What do you think? Could there be a better way to allow widget accessing data from collections other than its own? If there's already a plan for this feature in version 3, I think it'd make sense to return the payload in `loadEntry` for now so it doesn't break widgets that have already rely on this action.

Thank you again for the work here!

**Test plan**

If `loadEntry` is to stay in widget, I can add a test for it.

**A picture of a cute animal (not mandatory but encouraged)**

### CODE REVIEW

**Comment by erezrokah** [comment]
Thanks for this @d4rekanguok, looks like having a designated method for widgets is the way to go (and have it return a value).
I would even change it further so that method doesn't update any redux state, just an API style method to load an entry.
This might help avoid any future issues when the CMS changes.
Let make some attempts at this and I'll update soon.

**Comment by d4rekanguok** [comment]
Thank you for the quick response @erezrokah! That sounds great, please let me know if I can help in any capacity.

**Comment by erezrokah** [comment]
Another thing to consider is that in editorial workflow the `loadEntry` method is replaced with `loadUnpublishedEntry` here:
https://github.com/netlify/netlify-cms/blob/e81a35ccb8381c3132872f2d461134e1f4481080/packages/netlify-cms-core/src/components/Editor/withWorkflow.js#L34

**Comment by erezrokah** [comment]
@d4rekanguok, I pushed a new commit with another approach. Let me know what you think and would be great if you could test it (I tested it locally, but want to make sure it handles your use cases).

**Comment by d4rekanguok** [comment]
Thank you for this very quick turnaround! I've tested it out & spot a tiny bug where `tryLoadEntry` wasn't called with the right collection, other than that it worked great! 

Now that `tryLoadEntry` returns data directly, I can actually shorten my widget code a bit:
```diff,js
   const results = await loadEntry(collection, file)
-  const data = results.payload.entry.data[fieldName]
+  const data = results.data[fieldName]
```

- I notice that when a user pass in the wrong collection name & `tryLoadEntry` is called with a `undefined` collection, the error is not yet handled. It would be nice to be able to display this error in the editor UI, but I'm not sure how that can be done.

- This is outside the scope of this PR, but related: the action `query` is also passed to widget in a similar fashion to `loadEntry`. If this PR is to be merged, maybe `query` should be changed to use the same approach?
https://github.com/netlify/netlify-cms/blob/872e945c47ba5fd37fa608863864da96713ef7ad/packages/netlify-cms-core/src/actions/search.js#L149-L153

File: packages/netlify-cms-core/src/components/Editor/EditorControlPane/EditorControl.js [hunk 1]
```diff
@@ -270,7 +270,8 @@ const mapStateToProps = state => {
   const isLoadingAsset = selectIsLoadingAsset(state.medias);
 
   const loadEntry = (collectionName, slug) => {
-    return tryLoadEntry(state, collection, slug);
+    const targetCollection = collections.get(collectionName)
```
Comment (erezrokah) [inline packages/netlify-cms-core/src/components/Editor/EditorControlPane/EditorControl.js]: Thanks! Only tested with a single collection so I missed it...

**Comment by erezrokah** [comment]
> * the error is not yet handled. It would be nice to be able to display this error in the editor UI, but I'm not sure how that can be done.
> *

Added another commit to handle the error, I think the widget can catch it, store it in state and render something else on error. Not sure, but maybe even [using](https://www.netlifycms.org/docs/custom-widgets/#advanced-field-validation) to display the error.

**Comment by erezrokah** [comment]
> * This is outside the scope of this PR, but related: the action `query` is also passed to widget in a similar fashion to `loadEntry`. If this PR is to be merged, maybe `query` should be changed to use the same approach?

Makes sense, but probably in another PR as I think this will require updating the relation widget.

**Comment by d4rekanguok** [comment]
Awesome, thank you for resolving this so quickly!

**Pull request merged**